### ISSUE TITLE
prefer-destructuring function parameters

### ISSUE THREAD

**Issue by nickofthyme** [eslint/eslint#13383]
**What rule do you want to change?**
[`prefer-destructuring`](https://eslint.org/docs/rules/prefer-destructuring)

**Does this change cause the rule to produce more or fewer warnings?**
More

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option to add to second options with `enforceForRenamedProperties`, something like `enforceFunctionParameters`.

**Please provide some example code that this change will affect:**
Enforce destructuring on function parameters.

```js
// Bad
const data = [
  { foo: 1, bar: ['a'] },
  { foo: 2, bar: ['b'] },
  { foo: 3, bar: ['c'] },
];
const foos = data.map((d) => d.foo);
function getFoos(d) {
  return d.foo;
}
const foos = data.map((d) => d.bar[0]);
function getFoos(d) {
  return d.bar[0];
}

const getLines = (position) => ({
  x1: position[0],
  y1: position[1],
  x2: position[2],
  y2: position[3],
})
```

which fixing this would look something like...

```js
// Good
const data = [
  { foo: 1, bar: ['a'] },
  { foo: 2, bar: ['b'] },
  { foo: 3, bar: ['c'] },
];
const foos = data.map(({ foo }) => foo);
function getFoos({ foo }) {
  return foo;
}
const foos = data.map(({ bar: [first] }) => first);
function getFoos({ bar: [first] }) {
  return first;
}

const getLines = ([ x1, y1, x2, y2 ]) => ({
  x1,
  y1,
  x2,
  y2,
})
```

**What does the rule currently do for this code?**
Currently, none of the above examples would error on `prefer-destructuring`

**What will the rule do after it's changed?**
The same rules applied to object and array declarations and expressions would apply to function parameters that are objects or arrays.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by anikethsaha** [comment]
Thanks for the issue. 

IMO, this would need type information for the function parameters to know whether they can be destructible or not. 

May be this is a suitable enhancement for TS side, i.e. [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint)


let me know if I missed any

**Comment by nickofthyme** [comment]
Ok thanks. But couldn't you infer the type when each parameter is used.

```js
function getFoos(p1, p2) {
  return [p1.foo, p1];
}
```
In this case, the `p1` argument is not passed anywhere in the function as the object, only taking `foo` from the object. In such case, you can assume the destructuring of `p1` is allowable, right? And in this example, `p2` is not detructurable because we don't know the type and nothing in the signature tells us about the type.

I see it as those two checks to prefer destructuring
- Is the parameter used as an object in the function body?
- Is the parameter, the object, ever pass itself? Such that destructuring would not be advisable.

**Comment by kaicataldo** [comment]
The team has made the difficult decision to freeze stylistic rules (please see https://eslint.org/blog/2020/05/changes-to-rules-policies for more details), and we unfortunately won't be able to accept this proposal.
